// stdnum@1.4.10/lib/esm/cl/rut.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    const [v, err] = strings.cleanUnicode(input, ' -');
    if (err) {
        return ['', err];
    }
    if (v.startsWith('CL')) {
        return [v.substr(2), null];
    }
    return [v, null];
}
const impl = {
    name: 'Chilean National Tax Number',
    localName: 'Rol Único Tributario ',
    abbreviation: 'RUT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b, c, d] = strings.splitAt(value, 2, 5, 8);
        return `${a}.${b}.${c}-${d}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, check] = strings.splitAt(value, value.length - 1);
        if (!strings.isdigits(front)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = weightedSum(front, {
            reverse: true,
            weights: [9, 8, 7, 6, 5, 4, 9, 8, 7],
            modulus: 11,
        });
        const digit = '0123456789K'[sum];
        if (check !== digit) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
