// stdnum/lib/esm/id/npwp.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { luhnChecksumValidate } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
const NPWP_TAX_IDENTITIES = [
    '01',
    '02',
    '21',
    '31',
    '00',
    '20',
    '04',
    '05',
    '06',
    '07',
    '08',
    '09',
    '24',
    '25',
    '26',
    '31',
    '34',
    '35',
    '36',
    '47',
    '48',
    '49',
    '57',
    '58',
    '67',
    '77',
    '78',
    '79',
    '87',
    '88',
    '89',
    '97',
];
const impl = {
    name: 'Indonesian VAT Number',
    localName: 'Nomor Pokok Wajib Pajak',
    abbreviation: 'NPWP',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b, c, d, e, f] = strings.splitAt(value, 2, 5, 8, 9, 12);
        return `${a}.${b}.${c}.${d}-${e}.${f}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 15) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!NPWP_TAX_IDENTITIES.includes(value.substr(0, 2))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!luhnChecksumValidate(value.substr(0, 9))) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
