// stdnum@1.4.10/lib/esm/sv/nit.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('SV')) {
        return [value.substr(2), err];
    }
    return [value, err];
}
const impl = {
    name: 'El Salvador Tax Number',
    localName: 'Número de Identificación Tributaria',
    abbreviation: 'NIT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 4, -4, -1).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 14) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!['0', '1', '9'].includes(value[0])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        let sum;
        const [front, check] = strings.splitAt(value, 13);
        if (value.substr(10, 3) === '100') {
            sum =
                weightedSum(front, {
                    weights: [14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2],
                    modulus: 11,
                }) % 10;
        }
        else {
            sum =
                (11 -
                    weightedSum(front, {
                        weights: [2, 7, 6, 5, 4, 3, 2, 7, 6, 5, 4, 3, 2],
                        modulus: 11,
                    })) %
                    10;
        }
        if (check !== String(sum)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
