// stdnum@1.4.10/lib/esm/gt/nit.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const checkDigit = '0123456789K';
const impl = {
    name: 'Guatemala Tax Number',
    localName: 'Número de Identificación Tributaria',
    abbreviation: 'NIT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, -1).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length < 2 || value.length > 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, check] = strings.splitAt(value, -1);
        if (!strings.isdigits(front)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!checkDigit.includes(check)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = String((11 -
            weightedSum(front, {
                reverse: true,
                weights: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14],
                modulus: 11,
            })) %
            11);
        if (check !== checkDigit[sum]) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
