// stdnum@1.4.10/lib/cjs/cz/rc.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' /');
}
const impl = {
    name: 'Czech Birth Number',
    localName: 'Rodné číslo',
    abbreviation: 'RČ',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 6).join('/');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [yy, mm, dd] = util_1.strings.splitAt(value, 2, 4, 6);
        const mon = (parseInt(mm, 10) % 50) % 20;
        let year = parseInt(yy, 10) + 1900;
        if (value.length === 9) {
            if (year > 1980) {
                year -= 100;
            }
            if (year > 1953) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (year < 1954) {
            year += 100;
        }
        if (!(0, util_1.isValidDateCompactYYYYMMDD)(`${year}${String(mon).padStart(2, '0')}${dd}`)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (value.length === 10) {
            const [front, check] = util_1.strings.splitAt(value, -1);
            const sum = (parseInt(front, 10) % 11) % (year < 1985 ? 10 : 11);
            if (String(sum) !== check) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
