// stdnum@1.4.10/lib/cjs/de/stnr.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -./,');
}
function buildMatch(fmt) {
    const pattern = fmt.replace(/([FBUP])\1*/g, m => {
        return `(\\d{${m.length}})`;
    });
    const matcher = new RegExp(`^${pattern}$`);
    return (value) => {
        var _a, _b, _c, _d;
        const m = matcher.exec(value);
        if (!m) {
            return { match: false };
        }
        return {
            match: true,
            f: (_a = m[1]) !== null && _a !== void 0 ? _a : '',
            b: (_b = m[2]) !== null && _b !== void 0 ? _b : '',
            u: (_c = m[3]) !== null && _c !== void 0 ? _c : '',
            p: (_d = m[4]) !== null && _d !== void 0 ? _d : '',
        };
    };
}
function buildMatcher(rfmt, cfmt) {
    return {
        region: buildMatch(rfmt),
        country: buildMatch(cfmt),
    };
}
const REGION_FORMATS = {
    'DE-BW': buildMatcher('FFBBBUUUUP', '28FF0BBBUUUUP'),
    'DE-BY': buildMatcher('FFFBBBUUUUP', '9FFF0BBBUUUUP'),
    'DE-BE': buildMatcher('FFBBBUUUUP', '11FF0BBBUUUUP'),
    'DE-BB': buildMatcher('0FFBBBUUUUP', '30FF0BBBUUUUP'),
    'DE-HB': buildMatcher('FFBBBUUUUP', '24FF0BBBUUUUP'),
    'DE-HH': buildMatcher('FFBBBUUUUP', '22FF0BBBUUUUP'),
    'DE-HE': buildMatcher('0FFBBBUUUUP', '26FF0BBBUUUUP'),
    'DE-MV': buildMatcher('0FFBBBUUUUP', '40FF0BBBUUUUP'),
    'DE-NI': buildMatcher('FFBBBUUUUP', '23FF0BBBUUUUP'),
    'DE-NW': buildMatcher('FFFBBBBUUUP', '5FFF0BBBBUUUP'),
    'DE-RP': buildMatcher('FFBBBUUUUP', '27FF0BBBUUUUP'),
    'DE-SL': buildMatcher('0FFBBBUUUUP', '10FF0BBBUUUUP'),
    'DE-SN': buildMatcher('2FFBBBUUUUP', '32FF0BBBUUUUP'),
    'DE-ST': buildMatcher('1FFBBBUUUUP', '31FF0BBBUUUUP'),
    'DE-SH': buildMatcher('FFBBBUUUUP', '21FF0BBBUUUUP'),
    'DE-TH': buildMatcher('1FFBBBUUUUP', '41FF0BBBUUUUP'),
};
function findMatch(value) {
    let result = null;
    Object.values(REGION_FORMATS).some(({ region, country }) => {
        const rResult = region(value);
        if (rResult.match) {
            result = rResult;
            return true;
        }
        const cResult = country(value);
        if (cResult.match) {
            result = cResult;
            return true;
        }
        return false;
    });
    return result;
}
const impl = {
    name: 'German Tax Number',
    localName: 'Steuernummer',
    abbreviation: ' St.-Nr.',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const match = findMatch(input);
        if (!match || !match.match) {
            return value;
        }
        return `${match.f}/${match.b}/${match.u} ${match.p}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (![10, 11, 13].includes(value.length)) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (findMatch(value) === null) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const counter = {};
        value
            .substring(0, 10)
            .split('')
            .forEach(v => {
            var _a;
            counter[v] = ((_a = counter[v]) !== null && _a !== void 0 ? _a : 0) + 1;
        });
        const more = Object.values(counter);
        if (!more.some(v => v === 2 || v === 3)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
