// stdnum/lib/esm/gb/vat.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -/');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('GB')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
const impl = {
    name: 'United Kingdom (and Isle of Man) VAT Number',
    localName: 'Value Added Tax Registration Number',
    abbreviation: 'VAT Reg No',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 3, 7).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length === 5) {
            if (!strings.isdigits(value.substr(2))) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            const cvalue = parseInt(value.substr(2), 10);
            if (value.startsWith('GD') && cvalue >= 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (value.startsWith('HA') && cvalue < 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 11 &&
            (value.startsWith('GD8888') || value.startsWith('HA8888'))) {
            if (!strings.isdigits(value.substr(6))) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            const cvalue = parseInt(value.substr(6, 3), 10);
            if (value.startsWith('GD') && cvalue >= 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (value.startsWith('HA') && cvalue < 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (cvalue % 97 !== parseInt(value.substr(9, 2), 10)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if (value.length === 9 || value.length === 12) {
            if (!strings.isdigits(value)) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            const sum = weightedSum(value.substr(0, 9), {
                weights: [8, 7, 6, 5, 4, 3, 2, 10, 1],
                modulus: 97,
            });
            if (value[0] === '0') {
                if (![0, 42, 55].includes(sum)) {
                    return { isValid: false, error: new exceptions.InvalidChecksum() };
                }
            }
            else if (sum !== 0) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
