// stdnum@1.4.10/lib/esm/ar/cbu.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Single Banking Code',
    localName: 'Clave Bancaria Uniforme',
    abbreviation: 'CBU',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b] = strings.splitAt(value, 8);
        return `${a} ${b}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 22) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [front, c1, back, c2] = strings.splitAt(value, 7, 8, 21);
        const s1 = String(10 -
            weightedSum(front, {
                reverse: true,
                weights: [3, 1, 7, 9, 3, 1, 7],
                modulus: 10,
            }));
        const s2 = String(10 -
            weightedSum(back, {
                reverse: true,
                weights: [3, 1, 7, 9, 3, 1, 7, 9, 3, 1, 7, 9, 3, 1],
                modulus: 10,
            }));
        if (s1 !== c1 || s2 !== c2) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
