// stdnum@1.4.10/lib/esm/pt/nif.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -.', 'PT');
}
const impl = {
    name: 'Portuguese Tax Identification Number',
    localName: 'Número de Identificação Fiscal',
    abbreviation: 'NIF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 3, 6).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value) || value[0] === '0') {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, check] = strings.splitAt(value, -1);
        const sum = weightedSum(front, {
            weights: [9, 8, 7, 6, 5, 4, 3, 2, 1],
            modulus: 11,
        });
        if (String((11 - sum) % 10) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: '1234'.includes(value[0]),
            isCompany: !'1234'.includes(value[0]),
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
