// stdnum@1.4.10/lib/cjs/mx/curp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = exports.getBirthDate = exports.getGender = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' ');
}
const nameBlacklist = new Set([
    'BACA',
    'BAKA',
    'BUEI',
    'BUEY',
    'CACA',
    'CACO',
    'CAGA',
    'CAGO',
    'CAKA',
    'CAKO',
    'COGE',
    'COGI',
    'COJA',
    'COJE',
    'COJI',
    'COJO',
    'COLA',
    'CULO',
    'FALO',
    'FETO',
    'GETA',
    'GUEI',
    'GUEY',
    'JETA',
    'JOTO',
    'KACA',
    'KACO',
    'KAGA',
    'KAGO',
    'KAKA',
    'KAKO',
    'KOGE',
    'KOGI',
    'KOJA',
    'KOJE',
    'KOJI',
    'KOJO',
    'KOLA',
    'KULO',
    'LILO',
    'LOCA',
    'LOCO',
    'LOKA',
    'LOKO',
    'MAME',
    'MAMO',
    'MEAR',
    'MEAS',
    'MEON',
    'MIAR',
    'MION',
    'MOCO',
    'MOKO',
    'MULA',
    'MULO',
    'NACA',
    'NACO',
    'PEDA',
    'PEDO',
    'PENE',
    'PIPI',
    'PITO',
    'POPO',
    'PUTA',
    'PUTO',
    'QULO',
    'RATA',
    'ROBA',
    'ROBE',
    'ROBO',
    'RUIN',
    'SENO',
    'TETA',
    'VACA',
    'VAGA',
    'VAGO',
    'VAKA',
    'VUEI',
    'VUEY',
    'WUEI',
    'WUEY',
]);
const validStates = new Set([
    'AS',
    'BC',
    'BS',
    'CC',
    'CH',
    'CL',
    'CM',
    'CS',
    'DF',
    'DG',
    'GR',
    'GT',
    'HG',
    'JC',
    'MC',
    'MN',
    'MS',
    'NE',
    'NL',
    'NT',
    'OC',
    'PL',
    'QR',
    'QT',
    'SL',
    'SP',
    'SR',
    'TC',
    'TL',
    'TS',
    'VZ',
    'YN',
    'ZS',
]);
const checkAlphabet = '0123456789ABCDEFGHIJKLMN&OPQRSTUVWXYZ';
const checkAlphabetDict = checkAlphabet
    .split('')
    .reduce((acc, c, idx) => (Object.assign(Object.assign({}, acc), { [c]: idx })), {});
const impl = {
    name: 'Mexican Personal Identification',
    localName: 'Clave Única de Registro de Población',
    abbreviation: 'CURP',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 18) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!/^[A-Z]{4}[0-9]{6}[A-Z]{6}[0-9A-Z][0-9]$/.test(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!(0, util_1.isValidDateCompactYYMMDD)(value.substr(4, 6))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (nameBlacklist.has(value.substr(0, 4))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!['H', 'M'].includes(value[10])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!validStates.has(value.substr(11, 2))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const check = value
            .substr(0, 17)
            .split('')
            .reduce((acc, c, idx) => { var _a; return acc + ((_a = checkAlphabetDict[c]) !== null && _a !== void 0 ? _a : 0) * (18 - idx); }, 0);
        const checkStr = String((10 - (check % 10)) % 10);
        if (checkStr !== value.substr(17, 1)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
function getBirthDateImpl(value) {
    const parts = util_1.strings.splitAt(value, 4, 6, 8);
    const yyN = parseInt(parts[1], 10);
    const mmN = parseInt(parts[2], 10) - 1;
    const ddN = parseInt(parts[3], 10);
    if (!Number.isNaN(parseInt(value[16], 10))) {
        return new Date(yyN + 1900, mmN, ddN);
    }
    return new Date(yyN + 2000, mmN, ddN);
}
function getGender(input) {
    const value = impl.compact(input);
    return value[10] === 'H' ? 'M' : 'F';
}
exports.getGender = getGender;
function getBirthDate(input) {
    const value = impl.compact(input);
    return getBirthDateImpl(value);
}
exports.getBirthDate = getBirthDate;
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
