// stdnum@1.4.10/lib/cjs/dk/cpr.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = exports.getBirthDate = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function getBirthDate(value) {
    const [dob] = util_1.strings.splitAt(value, 6);
    let [day, month, year] = util_1.strings.splitAt(dob, 2, 4).map(v => Number(v));
    if ('5678'.includes(value[6]) && year >= 58) {
        year += 1800;
    }
    else if ('0123'.includes(value[6]) ||
        ('49'.includes(value[6]) && year >= 37)) {
        year += 1900;
    }
    else {
        year += 2000;
    }
    const date = new Date(year, month - 1, day);
    if (Number.isNaN(date.getTime()) ||
        [year, String(month).padStart(2, '0'), String(day).padStart(2, '0')].join('-') !== date.toISOString().substr(0, 10)) {
        throw new exceptions.InvalidComponent('The number does not contain valid birth date information.');
    }
    return date;
}
exports.getBirthDate = getBirthDate;
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -');
    if (err !== null) {
        return [value, err];
    }
    return [value, null];
}
const impl = {
    name: 'Danish Citizen Number',
    localName: 'Personnummer',
    abbreviation: 'CPR',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 6).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        try {
            const date = getBirthDate(value);
            if (date.getTime() > new Date().getTime()) {
                return {
                    isValid: false,
                    error: new exceptions.InvalidComponent('The birth date information is valid, but this person has not been born yet.'),
                };
            }
        }
        catch (err) {
            if (err instanceof exceptions.ValidationError) {
                return {
                    isValid: false,
                    error: err,
                };
            }
            return {
                isValid: false,
                error: new exceptions.InvalidComponent(String(err)),
            };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
