// stdnum/lib/esm/sm/coe.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
const lowNumbers = new Set([
    2, 4, 6, 7, 8, 9, 10, 11, 13, 16, 18, 19, 20, 21, 25, 26, 30, 32, 33, 35, 36,
    37, 38, 39, 40, 42, 45, 47, 49, 51, 52, 55, 56, 57, 58, 59, 61, 62, 64, 65,
    66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 79, 80, 81, 84, 85, 87, 88, 91,
    92, 94, 95, 96, 97, 99,
]);
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' .');
    if (err) {
        return [value, err];
    }
    return [value.replace(/^0+/, ''), null];
}
const impl = {
    name: 'San Marino National Tax Number',
    localName: 'Codice Operatore Eeconomico',
    abbreviation: 'COE',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length > 5 || value.length === 0) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length < 3 && !lowNumbers.has(parseInt(value, 10))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
