// stdnum/lib/esm/be/bis.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { validStructure, validChecksum, toDateArray } from './personIdentifierHelpers';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
function toDob(firstSix) {
    const [y, m, d] = toDateArray(firstSix).map(s => parseInt(s, 10));
    const adjustedDateArrays = [[y, m - 20, d], [y, m - 40, d]];
    const dobArray = adjustedDateArrays.find(ada => ada[1] >= 0 && ada[1] <= 12) || [];
    return dobArray.map(n => `${n}`.padStart(2, '0')).join('');
}
const impl = {
    name: 'Belgian Number for Foreigners',
    localName: 'Numéro BIS',
    abbreviation: 'BIS',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const number = impl.compact(input);
        if (!strings.isdigits(number) || parseInt(number, 10) <= 0) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (number.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!validStructure(number, toDob)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!validChecksum(number, toDob)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: number,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
