// stdnum/lib/esm/de/idnr.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { iso7064mod10x11validate } from '../util/iso7064';
function clean(input) {
    return strings.cleanUnicode(input, ' -./,');
}
const impl = {
    name: 'German Personal Tax Number',
    localName: 'Steuerliche Identifikationsnummer',
    abbreviation: 'IdNr',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 2, 5, 8).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value[0] === '0') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const counter = {};
        value
            .substr(0, 10)
            .split('')
            .map(v => parseInt(v, 10))
            .forEach(v => {
            var _a;
            counter[v] = ((_a = counter[v]) !== null && _a !== void 0 ? _a : 0) + 1;
        });
        const more = Object.values(counter).filter(v => v > 1);
        if (more.length !== 1 && [2, 3].includes(more[0])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!iso7064mod10x11validate(value)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
