// stdnum/lib/esm/de/vat.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { iso7064mod10x11validate } from '../util/iso7064';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -./,');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('DE')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
const impl = {
    name: 'German VAT Number',
    localName: 'Umsatzsteuer Identifikationsnummer',
    abbreviation: 'USt ID Nr.',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!iso7064mod10x11validate(value)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
