// stdnum/lib/esm/mx/rfc.js
import * as exceptions from '../exceptions';
import { isValidDateCompactYYMMDD, strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, '-_ ');
}
const nameBlacklist = new Set([
    'BUEI',
    'BUEY',
    'CACA',
    'CACO',
    'CAGA',
    'CAGO',
    'CAKA',
    'CAKO',
    'COGE',
    'COJA',
    'COJE',
    'COJI',
    'COJO',
    'CULO',
    'FETO',
    'GUEY',
    'JOTO',
    'KACA',
    'KACO',
    'KAGA',
    'KAGO',
    'KAKA',
    'KOGE',
    'KOJO',
    'KULO',
    'MAME',
    'MAMO',
    'MEAR',
    'MEAS',
    'MEON',
    'MION',
    'MOCO',
    'MULA',
    'PEDA',
    'PEDO',
    'PENE',
    'PUTA',
    'PUTO',
    'QULO',
    'RATA',
    'RUIN',
]);
const checkAlphabet = '0123456789ABCDEFGHIJKLMN&OPQRSTUVWXYZ Ñ';
const impl = {
    name: 'Mexican Tax Number',
    localName: 'Registro Federal de Contribuyentes',
    abbreviation: 'RFC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        if (value.length === 12) {
            return strings.splitAt(value, 3, 9).join(' ');
        }
        if (value.length === 13) {
            return strings.splitAt(value, 4, 10).join(' ');
        }
        return strings.splitAt(value, 4).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length === 10 || value.length === 13) {
            if (!/^[A-Z&Ñ]{4}[0-9]{6}([0-9A-Z]{3})?$/.test(value)) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (nameBlacklist.has(value.substr(0, 4))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!isValidDateCompactYYMMDD(value.substr(4, 6))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 12) {
            if (!/^[A-Z&Ñ]{3}[0-9]{6}[0-9A-Z]{3}$/.test(value)) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!isValidDateCompactYYMMDD(value.substr(3, 6))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (value.length >= 12) {
            if (!/[1-9A-V][1-9A-Z][0-9A]$/.test(value)) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            const [front, check] = strings.splitAt(value, -1);
            const sum = weightedSum(front.padStart(12, ' '), {
                modulus: 11,
                alphabet: checkAlphabet,
                weights: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14],
                reverse: true,
            });
            const mod = 11 - (sum % 11);
            let val;
            if (mod === 11) {
                val = '0';
            }
            else if (mod === 10) {
                val = 'A';
            }
            else {
                val = String(mod);
            }
            if (check !== val) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: value.length !== 12,
            isCompany: value.length === 12,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
