// stdnum/lib/esm/nl/btw.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { mod97base10Validate, weightedSum } from '../util/checksum';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -.', 'NL');
    if (err) {
        return [value, err];
    }
    const [a, b] = strings.splitAt(value, -3);
    return [`${a.padStart(9, '0')}${b}`, null];
}
const impl = {
    name: 'Dutch VAT Number',
    localName: 'Btw-identificatienummer',
    abbreviation: 'Btw-nr.',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [a, b, c] = strings.splitAt(value, 9, 10);
        if (!strings.isdigits(a) || !strings.isdigits(c)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (b !== 'B') {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const sum = weightedSum(a, {
            weights: [9, 8, 7, 6, 5, 4, 3, 2, -1],
            modulus: 10000,
        });
        if (sum % 11 !== 0 && !mod97base10Validate(`NL${a}${b}${c}`)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: `${a}${b}${c}`,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
