// stdnum@1.4.10/lib/esm/kr/rrn.js
import * as exceptions from '../exceptions';
import { isValidDateCompactYYYYMMDD, strings } from '../util';
import { weightedSum } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -/');
}
const centuryPrefix = {
    0: 18,
    1: 19,
    2: 19,
    3: 20,
    4: 20,
    5: 19,
    6: 19,
    7: 20,
    8: 20,
    9: 18,
};
const impl = {
    name: 'South Korean Resident Registration Number',
    localName: '주민등록번호',
    abbreviation: 'RRN',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 6).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 13) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [dob, century, place, , check] = strings.splitAt(value, 6, 7, 9, 12);
        if (!isValidDateCompactYYYYMMDD(`${centuryPrefix[century]}${dob}`)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (parseInt(place, 10) > 96) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = weightedSum(value.substr(0, 12), {
            modulus: 11,
            weights: [2, 3, 4, 5, 6, 7, 8, 9, 2, 3, 4, 5],
        });
        if (String((11 - sum) % 10) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
