// stdnum@1.4.10/lib/esm/cr/cpf.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' ');
    if (err) {
        return [value, err];
    }
    const parts = value.split('-');
    if (parts.length === 3) {
        parts[0] = parts[0].padStart(2, '0');
        parts[1] = parts[1].padStart(4, '0');
        parts[2] = parts[2].padStart(4, '0');
    }
    const number = parts.join('');
    if (number.length !== 10 && number.length !== 9) {
        return ['', new exceptions.InvalidLength()];
    }
    return [number.padStart(10, '0'), err];
}
const impl = {
    name: 'Costa Rica Physical Person ID Number',
    localName: 'Cédula de Persona Física',
    abbreviation: 'CPF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 2, 6).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value[0] !== '0') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
