// stdnum@1.4.10/lib/cjs/ie/vat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = exports.calcCheckDigit = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' ', 'IE');
}
const ALPHABET = 'WABCDEFGHIJKLMNOPQRSTUV';
function calcCheckDigit(valueIn) {
    const value = valueIn.padStart(7, '0');
    const d1 = value
        .substr(0, 7)
        .split('')
        .reduce((acc, vv, idx) => {
        const v = Number.parseInt(vv, 10);
        return acc + (8 - idx) * v;
    }, 0);
    const d2 = value.length === 8 ? 9 * ALPHABET.indexOf(value[7]) : 0;
    return ALPHABET[(d1 + d2) % 23];
}
exports.calcCheckDigit = calcCheckDigit;
const impl = {
    name: 'Irish VAT Number',
    localName: 'Cáin Bhreisluacha',
    abbreviation: 'CBL',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value[0]) || !util_1.strings.isdigits(value.substr(2, 5))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, end] = util_1.strings.splitAt(value, 7);
        if (!end.split('').every(v => ALPHABET.includes(v))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (util_1.strings.isdigits(front)) {
            if (value[7] !== calcCheckDigit(`${front}${value.substr(8)}`)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if ('ABCDEFGHIJKLMNOPQRSTUVWXYZ+*'.includes(value[1])) {
            if (value[7] !== calcCheckDigit(`${value.substr(2, 5)}${value[0]}`)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
