// stdnum/lib/esm/ve/rif.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
const companyTypes = {
    V: 4,
    E: 8,
    J: 12,
    P: 16,
    G: 20,
};
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Venezuelan VAT Number',
    localName: 'Registro Único de Información Fiscal',
    abbreviation: 'RIF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1, 9).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value.substr(1))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [ctype, body, check] = strings.splitAt(value, 1, 9);
        const first = companyTypes[ctype];
        if (first === undefined) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const digit = (first +
            weightedSum(body, {
                weights: [3, 2, 7, 6, 5, 4, 3, 2],
                modulus: 11,
            })) %
            11;
        if ('00987654321'[digit] !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: ['V', 'E'].includes(ctype),
            isCompany: ['J', 'G', 'P'].includes(ctype),
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
