// stdnum/lib/esm/nz/ird.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { weightedSum } from '../util/checksum';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('NZ')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
const impl = {
    name: 'Inland Revenue Department Number',
    localName: 'Te Tari Taake',
    abbreviation: 'IRD',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 2, 5).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [frontV, check] = strings.splitAt(value, -1);
        const front = frontV.padStart(8, '0');
        let sum = (11 -
            weightedSum(front, {
                weights: [3, 2, 7, 6, 5, 4, 3, 2],
                modulus: 11,
            })) %
            11;
        if (sum === 10) {
            sum =
                (11 -
                    weightedSum(front, {
                        weights: [7, 4, 3, 2, 5, 2, 7, 6],
                        modulus: 11,
                    })) %
                    11;
        }
        if (String(sum) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
