// stdnum@1.4.10/lib/esm/cu/ni.js
import * as exceptions from '../exceptions';
import { isValidDateCompactYYYYMMDD, strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'Cuban Identity Card Number',
    localName: 'Número de Identidad',
    abbreviation: 'NI',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [yy, mm, dd, cen] = strings.splitAt(value, 2, 4, 6, 7);
        let year;
        if ('012345'.includes(cen)) {
            year = `19${yy}`;
        }
        else if ('678'.includes(cen)) {
            year = `20${yy}`;
        }
        else if (cen === '9') {
            year = `18${yy}`;
        }
        else {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!isValidDateCompactYYYYMMDD(`${year}${mm}${dd}`)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
