// stdnum/lib/esm/ie/pps.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { calcCheckDigit } from './vat';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const ppsRe = /^\d{7}[A-W][AHWTX]?$/;
const impl = {
    name: 'Irish Personal Number',
    localName: 'Personal Public Service Number',
    abbreviation: 'PPS',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!ppsRe.test(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length === 9 && 'AH'.includes(value[8])) {
            if (value[7] !== calcCheckDigit(`${value.substr(0, 7)}${value.substr(8)}`)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if (value[7] !== calcCheckDigit(value.substr(0, 7))) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
