// stdnum/lib/esm/ie/vat.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' ', 'IE');
}
const ALPHABET = 'WABCDEFGHIJKLMNOPQRSTUV';
export function calcCheckDigit(valueIn) {
    const value = valueIn.padStart(7, '0');
    const d1 = value
        .substr(0, 7)
        .split('')
        .reduce((acc, vv, idx) => {
        const v = Number.parseInt(vv, 10);
        return acc + (8 - idx) * v;
    }, 0);
    const d2 = value.length === 8 ? 9 * ALPHABET.indexOf(value[7]) : 0;
    return ALPHABET[(d1 + d2) % 23];
}
const impl = {
    name: 'Irish VAT Number',
    localName: 'Cáin Bhreisluacha',
    abbreviation: 'CBL',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value[0]) || !strings.isdigits(value.substr(2, 5))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, end] = strings.splitAt(value, 7);
        if (!end.split('').every(v => ALPHABET.includes(v))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (strings.isdigits(front)) {
            if (value[7] !== calcCheckDigit(`${front}${value.substr(8)}`)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if ('ABCDEFGHIJKLMNOPQRSTUVWXYZ+*'.includes(value[1])) {
            if (value[7] !== calcCheckDigit(`${value.substr(2, 5)}${value[0]}`)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
