// stdnum@1.4.10/lib/cjs/it/aic.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' ');
}
const alphabet = '0123456789BCDFGHJKLMNPQRSTUVWXYZ';
const impl = {
    name: 'Italian Code for Identification of Drugs',
    localName: 'Autorizzazione all’Immissione in Commercio',
    abbreviation: 'AIC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 6 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        let base10 = value;
        if (value.length === 6) {
            const chars = value.split('');
            if (!chars.every(v => alphabet.includes(v))) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            base10 = String(chars
                .reverse()
                .reduce((acc, v, idx) => acc + alphabet.indexOf(v) * Math.pow(32, idx), 0)).padStart(9, '0');
        }
        if (!util_1.strings.isdigits(base10)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, end] = util_1.strings.splitAt(base10, -1);
        const sum = front
            .split('')
            .map(v => parseInt(v, 10))
            .reduce((acc, v, idx) => {
            const vv = v * ((idx % 2) + 1);
            return acc + Math.floor(vv / 10) + (vv % 10);
        }, 0);
        if (String(sum % 10) !== end) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
