// stdnum/lib/esm/us/ein.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
const prefixes = [
    '01',
    '02',
    '03',
    '04',
    '05',
    '06',
    '11',
    '13',
    '14',
    '16',
    '21',
    '22',
    '23',
    '25',
    '34',
    '51',
    '52',
    '54',
    '55',
    '56',
    '57',
    '58',
    '59',
    '65',
    '10',
    '12',
    '15',
    '24',
    '20',
    '26',
    '27',
    '45',
    '46',
    '47',
    '81',
    '82',
    '83',
    '84',
    '30',
    '32',
    '35',
    '36',
    '37',
    '38',
    '61',
    '31',
    '33',
    '39',
    '41',
    '42',
    '43',
    '46',
    '48',
    '62',
    '63',
    '64',
    '66',
    '68',
    '71',
    '72',
    '73',
    '74',
    '75',
    '76',
    '77',
    '85',
    '86',
    '87',
    '88',
    '91',
    '92',
    '93',
    '98',
    '99',
    '40',
    '44',
    '50',
    '53',
    '60',
    '67',
    '80',
    '90',
    '94',
    '95',
];
function clean(input) {
    return strings.cleanUnicode(input, '- ');
}
const impl = {
    name: 'U.S. Employer Identification Number',
    localName: 'Employer Identification Number',
    abbreviation: 'EIN',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 2).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!prefixes.includes(value.substr(0, 2))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
