// stdnum/lib/cjs/uy/rut.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const pymod_1 = require("../util/pymod");
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -');
    if (err) {
        return [value, err];
    }
    if (value.startsWith('UY')) {
        return [value.substr(2), null];
    }
    return [value, err];
}
const impl = {
    name: 'Uruguayan Tax Number',
    localName: 'Registro Único Tributario',
    abbreviation: 'RUT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 2, -4, -1).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (parseInt(value.substr(0, 2), 10) > 21 || value.substr(0, 2) === '00') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (value.substr(2, 6) === '000000') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (value.substr(8, 3) !== '001') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [front, check] = util_1.strings.splitAt(value, 11);
        const sum = (0, util_1.weightedSum)(front, {
            weights: [4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2],
            modulus: 11,
        });
        const digit = String((0, pymod_1.pymod)(-sum, 11));
        if (check !== digit) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
