// stdnum/lib/esm/cn/ric.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
import { isValidDateCompactYYYYMMDD } from '../util/isValidDate';
function clean(input) {
    return strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'Chinese Resident Identity Card Number',
    localName: '居民身份证',
    abbreviation: 'RIC No',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 18) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, check] = strings.splitAt(value, 17);
        if (!strings.isdigits(front)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!isValidDateCompactYYYYMMDD(front.substr(6, 8))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const checkValue = weightedSum(front, {
            weights: [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2],
            modulus: 11,
        });
        const digit = '10X98765432'[checkValue];
        if (check !== digit) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
