// stdnum/lib/esm/cn/uscc.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const alphabet = '0123456789ABCDEFGHJKLMNPQRTUWXY';
const impl = {
    name: 'Chinese Unified Social Credit Code',
    localName: '统一社会信用代码',
    abbreviation: 'USCC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 18) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, back, check] = strings.splitAt(value, 8, 17);
        if (!strings.isdigits(front)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (back.split('').some(v => !alphabet.includes(v))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const digit = weightedSum(value.substr(0, 17), {
            weights: [1, 3, 9, 27, 19, 26, 16, 17, 20, 29, 25, 13, 8, 24, 10, 30, 28],
            modulus: 31,
            alphabet,
        });
        if (alphabet[31 - digit] !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
