// stdnum@1.4.10/lib/esm/vn/mst.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { weightedSum } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Vietnamese Tax Number',
    localName: 'Mã số thuế',
    abbreviation: 'MST',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 10).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10 && value.length !== 13) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [province, sequence, check, suffix] = strings.splitAt(value, 2, 9, 10);
        if (sequence === '0000000' || suffix === '000' || province === '00') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = weightedSum(value.substr(0, 9), {
            modulus: 11,
            weights: [31, 29, 23, 19, 17, 13, 7, 5, 3],
        });
        if (check !== String(10 - sum)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
