// stdnum@1.4.10/lib/esm/ca/bn.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { luhnChecksumValidate } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Canadian Business Number',
    localName: 'Business Number',
    abbreviation: 'BN',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        if (value.length === 15) {
            return strings.splitAt(value, 5, 9, 11).join(' ');
        }
        return strings.splitAt(value, 5).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9 && value.length !== 15) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, rest] = strings.splitAt(value, 9);
        if (!strings.isdigits(front)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!luhnChecksumValidate(front)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        if (rest) {
            const [a, b] = strings.splitAt(rest, 2);
            if (!['RC', 'RM', 'RP', 'RT'].includes(a)) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!strings.isdigits(b)) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
