// stdnum/lib/cjs/ch/vat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const uid_1 = require("./uid");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Swiss VAT Number',
    localName: 'Mehrwertsteuernummer',
    abbreviation: 'MWST/TVA/IVA',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b, c, d, e] = util_1.strings.splitAt(value, 3, 6, 9, 12);
        return `${a}-${b}.${c}.${d} ${e}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 15 && value.length !== 16) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, sufix] = util_1.strings.splitAt(value, 12);
        if (!['MWST', 'TVA', 'IVA', 'TPV'].includes(sufix)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const result = (0, uid_1.validate)(front);
        if (!result.isValid && result.error) {
            return { isValid: false, error: result.error };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
