// stdnum@1.4.10/lib/esm/ch/uid.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { weightedSum } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Swiss Business Identifier',
    localName: 'Unternehmens-Identifikationsnummer',
    abbreviation: 'UID',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b, c, d] = strings.splitAt(value, 3, 6, 9);
        return `${a}-${b}.${c}.${d}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value.substr(3)) || !value.startsWith('CHE')) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [, front, check] = strings.splitAt(value, 3, -1);
        const sum = weightedSum(front, {
            modulus: 11,
            weights: [5, 4, 3, 2, 7, 6, 5, 4],
        });
        if (String((11 - sum) % 11) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
