// stdnum@1.4.10/lib/esm/ch/vat.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { validate as uidValidate } from './uid';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Swiss VAT Number',
    localName: 'Mehrwertsteuernummer',
    abbreviation: 'MWST/TVA/IVA',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b, c, d, e] = strings.splitAt(value, 3, 6, 9, 12);
        return `${a}-${b}.${c}.${d} ${e}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 15 && value.length !== 16) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, sufix] = strings.splitAt(value, 12);
        if (!['MWST', 'TVA', 'IVA', 'TPV'].includes(sufix)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const result = uidValidate(front);
        if (!result.isValid && result.error) {
            return { isValid: false, error: result.error };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
