// stdnum/lib/esm/ru/inn.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'Russian Tax Identifier',
    localName: 'Идентификационный номер налогоплательщика',
    abbreviation: 'ИНН',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10 && value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        let digit;
        const [front, check] = strings.splitAt(value, value.length === 10 ? -1 : -2);
        if (value.length === 10) {
            digit = String(weightedSum(front, {
                weights: [2, 4, 10, 3, 5, 9, 4, 6, 8],
                modulus: 11,
            }) % 10);
        }
        else {
            const d1 = String(weightedSum(front, {
                weights: [7, 2, 4, 10, 3, 5, 9, 4, 6, 8],
                modulus: 11,
            }) % 10);
            const d2 = String(weightedSum(front + d1, {
                weights: [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8],
                modulus: 11,
            }) % 10);
            digit = `${d1}${d2}`;
        }
        if (digit !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: value.length === 12,
            isCompany: value.length === 10,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
