// stdnum@1.4.10/lib/cjs/ad/nrt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Andorra Tax Register Number',
    localName: 'Número de Registre Tributari',
    abbreviation: 'NRT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 1, 7).join('-');
    },
    validate(value) {
        const [v, error] = clean(value);
        if (error) {
            return { isValid: false, error };
        }
        if (v.length !== 8) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const mid = v.substr(1, v.length - 2);
        if (!util_1.strings.isalpha(v[0]) || !util_1.strings.isalpha(v[v.length - 1])) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!util_1.strings.isdigits(mid)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!'ACDEFGLOPU'.includes(v[0])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (v[0] === 'F' && mid > '699999') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if ('AL'.includes(v[0]) && mid > '699999' && mid < '800000') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: v,
            isIndividual: 'FE'.includes(v[0]),
            isCompany: !'FE'.includes(v[0]),
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
