// stdnum/lib/cjs/cz/dic.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const rc_1 = require("./rc");
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('CZ')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
function checkLegal(value) {
    const [front, check] = util_1.strings.splitAt(value, -1);
    const sum = (0, util_1.weightedSum)(front, {
        modulus: 11,
        weights: [8, 7, 6, 5, 4, 3, 2, 1],
    });
    const v = (11 - sum) % 11;
    return (v === 0 ? '1' : String(v % 10)) === check;
}
function checkSpecial(value) {
    const [front, check] = util_1.strings.splitAt(value, -1);
    const sum = (0, util_1.weightedSum)(front, {
        modulus: 11,
        weights: [8, 7, 6, 5, 4, 3, 2, 1],
    });
    const digit = String((8 - ((10 - sum) % 11)) % 10);
    return digit === check;
}
const impl = {
    name: 'Czech VAT Number',
    localName: 'Daňové identifikační číslo',
    abbreviation: 'DIČ',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (![8, 9, 10].includes(value.length)) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length === 8) {
            if (value.startsWith('9')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!checkLegal(value)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
            return {
                isValid: true,
                compact: value,
                isIndividual: false,
                isCompany: true,
            };
        }
        if (value.length === 9 && value.startsWith('6')) {
            if (!checkSpecial(value)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
            return {
                isValid: true,
                compact: value,
                isIndividual: false,
                isCompany: true,
            };
        }
        return (0, rc_1.validate)(value);
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
