// stdnum@1.4.10/lib/esm/cz/dic.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
import { validate as rcValidate } from './rc';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('CZ')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
function checkLegal(value) {
    const [front, check] = strings.splitAt(value, -1);
    const sum = weightedSum(front, {
        modulus: 11,
        weights: [8, 7, 6, 5, 4, 3, 2, 1],
    });
    const v = (11 - sum) % 11;
    return (v === 0 ? '1' : String(v % 10)) === check;
}
function checkSpecial(value) {
    const [front, check] = strings.splitAt(value, -1);
    const sum = weightedSum(front, {
        modulus: 11,
        weights: [8, 7, 6, 5, 4, 3, 2, 1],
    });
    const digit = String((8 - ((10 - sum) % 11)) % 10);
    return digit === check;
}
const impl = {
    name: 'Czech VAT Number',
    localName: 'Daňové identifikační číslo',
    abbreviation: 'DIČ',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (![8, 9, 10].includes(value.length)) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length === 8) {
            if (value.startsWith('9')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!checkLegal(value)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
            return {
                isValid: true,
                compact: value,
                isIndividual: false,
                isCompany: true,
            };
        }
        if (value.length === 9 && value.startsWith('6')) {
            if (!checkSpecial(value)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
            return {
                isValid: true,
                compact: value,
                isIndividual: false,
                isCompany: true,
            };
        }
        return rcValidate(value);
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
