// stdnum/lib/cjs/de/svnr.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const checksum_1 = require("../util/checksum");
const AREA_NUMBER_OPTIONS = [
    '02',
    '03',
    '04',
    '08',
    '09',
    '10',
    '11',
    '12',
    '13',
    '14',
    '15',
    '16',
    '17',
    '18',
    '19',
    '20',
    '21',
    '23',
    '24',
    '25',
    '26',
    '28',
    '29',
    '38',
    '39',
    '40',
    '80',
    '81',
    '82',
    '89',
].concat([...Array(79 - 42 + 1).keys()].map(x => (x + 42).toString()));
const BIRTH_MONTH_OPTIONS = [
    '01',
    '02',
    '03',
    '04',
    '05',
    '06',
    '07',
    '08',
    '09',
    '10',
    '11',
    '12',
];
const checkAlphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const checkAlphabetDict = checkAlphabet
    .split('')
    .reduce((acc, c, idx) => (Object.assign(Object.assign({}, acc), { [c]: String(idx + 1).padStart(2, '0') })), {});
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -./,');
}
const impl = {
    name: 'German Pension Insurance Number',
    localName: 'Sozialversicherungsnummer',
    abbreviation: 'SVNR',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 2, 8, 9).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isalphanumeric(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!AREA_NUMBER_OPTIONS.includes(value.substring(0, 2))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!util_1.strings.isdigits(value.substring(2, 4))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!BIRTH_MONTH_OPTIONS.includes(value.substring(4, 6))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!util_1.strings.isdigits(value.substring(6, 8))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!util_1.strings.isalpha(value[8])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!util_1.strings.isdigits(value.substring(9, 11))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [frontWithAlpha, check] = util_1.strings.splitAt(value, 11);
        const front = frontWithAlpha
            .split('')
            .map(c => { var _a; return (_a = checkAlphabetDict[c]) !== null && _a !== void 0 ? _a : c; })
            .join('');
        const sum = (0, checksum_1.weightedSum)(front, {
            weights: [2, 1, 2, 5, 7, 1, 2, 1, 2, 1, 2, 1],
            modulus: 10,
            sumByDigit: true,
        });
        if (String(sum % 10) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
