// stdnum@1.4.10/lib/esm/hk/hkid.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { weightedSum } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -()');
}
const alphabet = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const impl = {
    name: 'Hong Kong Identity Card Number',
    localName: '香港身份證',
    abbreviation: 'HKID',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!/^[A-NP-Z]{1,2}[0-9]{6}[0-9A]$/.test(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, check] = strings.splitAt(value, -1);
        const sum = (weightedSum(front, {
            modulus: 11,
            weights: front.length === 7
                ? [8, 7, 6, 5, 4, 3, 2, 1]
                : [9, 8, 7, 6, 5, 4, 3, 2, 1],
            alphabet,
        }) +
            (front.length === 7 ? 5 : 0)) %
            11;
        if (alphabet[(11 - sum) % 11] !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
