// stdnum/lib/esm/kr/brn.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { weightedSum } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'South Korean Business Registration Number',
    localName: '사업자 등록 번호',
    abbreviation: 'BRN',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 3, 5).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [head, mid, end, check] = strings.splitAt(value, 3, 5, 9);
        if (parseInt(head, 10) < 101) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (mid === '00') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (end === '0000') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = weightedSum(value.substr(0, 9), {
            modulus: 10,
            weights: [1, 3, 7, 1, 3, 7, 1, 3, 5],
        });
        const extra = Math.floor((parseInt(end[3], 10) * 5) / 10);
        if (String((10 - ((sum + extra) % 10)) % 10) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
