// stdnum@1.4.10/lib/cjs/do/ncf.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const ncfTypes = [
    '01',
    '02',
    '03',
    '04',
    '11',
    '12',
    '13',
    '14',
    '15',
    '16',
    '17',
];
const ecfTypes = [
    '31',
    '32',
    '33',
    '34',
    '41',
    '43',
    '44',
    '45',
];
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'Dominican Republic Receipt Number',
    localName: 'Números de Comprobante Fiscal',
    abbreviation: 'NCF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length === 13) {
            if (!value.startsWith('E')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!util_1.strings.isdigits(value.substr(1))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!ecfTypes.includes(value.substr(1, 2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 11) {
            if (!value.startsWith('B')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!util_1.strings.isdigits(value.substr(1))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!ncfTypes.includes(value.substr(1, 2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 19) {
            if (!value.startsWith('AP')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!util_1.strings.isdigits(value.substr(2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!ncfTypes.includes(value.substr(9, 2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
