// stdnum/lib/esm/my/nric.js
import * as exceptions from '../exceptions';
import { isValidDateCompactYYMMDD, strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -*');
}
const UNASSIGNED = [
    '00',
    '17',
    '18',
    '19',
    '20',
    '69',
    '73',
    '80',
    '81',
    '94',
    '95',
    '96',
    '97',
];
const impl = {
    name: 'Malaysian National Registration Identity Card Number',
    localName: 'National Registration Identity Card Number',
    abbreviation: 'NRIC No.',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 6, 8).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [bdate, place] = strings.splitAt(value, 6, 8);
        if (!isValidDateCompactYYMMDD(bdate)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (UNASSIGNED.includes(place)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
