// stdnum@1.4.10/lib/cjs/fr/tva.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const siren = __importStar(require("./siren"));
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -.');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('FR')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
const alphabet = '0123456789ABCDEFGHJKLMNPQRSTUVWXYZ';
const impl = {
    name: 'French VAT Number',
    localName: "Numéro d'Identification à la Taxe sur la Valeur Ajoutée",
    abbreviation: 'n°TVA',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 2, 5, 8).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!alphabet.includes(value[0]) || !alphabet.includes(value[1])) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!util_1.strings.isdigits(value.substr(2))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.substr(2, 3) !== '000') {
            const r = siren.validate(value.substr(2));
            if (!r.isValid) {
                return r;
            }
        }
        if (util_1.strings.isdigits(value)) {
            const [check, back] = util_1.strings.splitAt(value, 2);
            const sum = parseInt(`${back}12`, 10) % 97;
            if (parseInt(check, 10) !== sum) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            const [check, back] = util_1.strings.splitAt(value, 2);
            let cvalue;
            if (util_1.strings.isdigits(check[0])) {
                cvalue =
                    alphabet.indexOf(check[0]) * 24 + alphabet.indexOf(check[1]) - 10;
            }
            else {
                cvalue =
                    alphabet.indexOf(check[0]) * 34 + alphabet.indexOf(check[1]) - 100;
            }
            const sum = (parseInt(back, 10) + 1 + Math.floor(cvalue / 11)) % 11;
            const digit = cvalue % 11;
            if (sum !== digit) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
