// stdnum@1.4.10/lib/esm/do/ncf.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
const ncfTypes = [
    '01',
    '02',
    '03',
    '04',
    '11',
    '12',
    '13',
    '14',
    '15',
    '16',
    '17',
];
const ecfTypes = [
    '31',
    '32',
    '33',
    '34',
    '41',
    '43',
    '44',
    '45',
];
function clean(input) {
    return strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'Dominican Republic Receipt Number',
    localName: 'Números de Comprobante Fiscal',
    abbreviation: 'NCF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length === 13) {
            if (!value.startsWith('E')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!strings.isdigits(value.substr(1))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!ecfTypes.includes(value.substr(1, 2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 11) {
            if (!value.startsWith('B')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!strings.isdigits(value.substr(1))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!ncfTypes.includes(value.substr(1, 2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 19) {
            if (!value.startsWith('AP')) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!strings.isdigits(value.substr(2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (!ncfTypes.includes(value.substr(9, 2))) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
