// stdnum/lib/cjs/fi/hetu.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' ');
}
const CHECK_ALPHA = '0123456789ABCDEFHJKLMNPRSTUVWXY';
const CENTURY = {
    '+': '18',
    '-': '19',
    A: '20',
};
const impl = {
    name: 'Finnish Personal Identity Code',
    localName: 'Henkilötunnus',
    abbreviation: 'HETU',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [front, , back, check2] = util_1.strings.splitAt(value, 6, 7, -1);
        if (!util_1.strings.isdigits(front) ||
            !util_1.strings.isdigits(back) ||
            !CHECK_ALPHA.includes(check2)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!'-+A'.includes(value[6])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [dd, mm, yy, century, person, check] = util_1.strings.splitAt(value, 2, 4, 6, 7, 10);
        const dstr = `${CENTURY[century]}${yy}${mm}${dd}`;
        if (!(0, util_1.isValidDateCompactYYYYMMDD)(dstr)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const checkable = `${dd}${mm}${yy}${person}`;
        if (CHECK_ALPHA[parseInt(checkable, 10) % CHECK_ALPHA.length] !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
