// stdnum@1.4.10/lib/esm/it/aic.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' ');
}
const alphabet = '0123456789BCDFGHJKLMNPQRSTUVWXYZ';
const impl = {
    name: 'Italian Code for Identification of Drugs',
    localName: 'Autorizzazione all’Immissione in Commercio',
    abbreviation: 'AIC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 6 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        let base10 = value;
        if (value.length === 6) {
            const chars = value.split('');
            if (!chars.every(v => alphabet.includes(v))) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            base10 = String(chars
                .reverse()
                .reduce((acc, v, idx) => acc + alphabet.indexOf(v) * Math.pow(32, idx), 0)).padStart(9, '0');
        }
        if (!strings.isdigits(base10)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, end] = strings.splitAt(base10, -1);
        const sum = front
            .split('')
            .map(v => parseInt(v, 10))
            .reduce((acc, v, idx) => {
            const vv = v * ((idx % 2) + 1);
            return acc + Math.floor(vv / 10) + (vv % 10);
        }, 0);
        if (String(sum % 10) !== end) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
