// stdnum@1.4.10/lib/esm/br/cpf.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
function computeDigit(input) {
    const mlen = input.length + 1;
    const value = input
        .split('')
        .map((v, idx) => parseInt(v, 10) * (mlen - idx))
        .reduce((acc, v) => (acc + v) % 11);
    return value < 2 ? 0 : 11 - value;
}
const impl = {
    name: 'Brazilian National Identifier',
    localName: 'Cadastro de Pessoas Físicas',
    abbreviation: 'CPF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        const [a, b, c, d] = strings.splitAt(value, 3, 6, input.length - 2);
        return `${a}.${b}.${c}-${d}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, c1, c2] = strings.splitAt(value, 9, 10);
        const d1 = String(computeDigit(front));
        const d2 = String(computeDigit(value.substr(0, 10)));
        if (d1 !== c1 || d2 !== c2) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
