// stdnum@1.4.10/lib/esm/uy/rut.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
import { pymod } from '../util/pymod';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -');
    if (err) {
        return [value, err];
    }
    if (value.startsWith('UY')) {
        return [value.substr(2), null];
    }
    return [value, err];
}
const impl = {
    name: 'Uruguayan Tax Number',
    localName: 'Registro Único Tributario',
    abbreviation: 'RUT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 2, -4, -1).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (parseInt(value.substr(0, 2), 10) > 21 || value.substr(0, 2) === '00') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (value.substr(2, 6) === '000000') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (value.substr(8, 3) !== '001') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [front, check] = strings.splitAt(value, 11);
        const sum = weightedSum(front, {
            weights: [4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2],
            modulus: 11,
        });
        const digit = String(pymod(-sum, 11));
        if (check !== digit) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
