// stdnum/lib/esm/cz/rc.js
import * as exceptions from '../exceptions';
import { isValidDateCompactYYYYMMDD, strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' /');
}
const impl = {
    name: 'Czech Birth Number',
    localName: 'Rodné číslo',
    abbreviation: 'RČ',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 6).join('/');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [yy, mm, dd] = strings.splitAt(value, 2, 4, 6);
        const mon = (parseInt(mm, 10) % 50) % 20;
        let year = parseInt(yy, 10) + 1900;
        if (value.length === 9) {
            if (year > 1980) {
                year -= 100;
            }
            if (year > 1953) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (year < 1954) {
            year += 100;
        }
        if (!isValidDateCompactYYYYMMDD(`${year}${String(mon).padStart(2, '0')}${dd}`)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (value.length === 10) {
            const [front, check] = strings.splitAt(value, -1);
            const sum = (parseInt(front, 10) % 11) % (year < 1985 ? 10 : 11);
            if (String(sum) !== check) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
