// stdnum@1.4.10/lib/cjs/bg/vat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const egn = __importStar(require("./egn"));
const pnf = __importStar(require("./pnf"));
const util_1 = require("../util");
const checksum_1 = require("../util/checksum");
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -/');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('BG')) {
        return [value.substr(2), null];
    }
    return [value, err];
}
function checkLegal(value) {
    const [front, check] = util_1.strings.splitAt(value, -1);
    let sum = (0, checksum_1.weightedSum)(front, {
        modulus: 11,
        weights: [1, 2, 3, 4, 5, 6, 7, 8],
    });
    if (sum === 10) {
        sum = (0, checksum_1.weightedSum)(front, {
            modulus: 11,
            weights: [3, 4, 5, 6, 7, 8, 9, 10],
        });
    }
    return String(sum % 10) === check;
}
function checkOther(value) {
    const [front, check] = util_1.strings.splitAt(value, -1);
    const sum = 11 -
        (0, checksum_1.weightedSum)(front, {
            modulus: 11,
            weights: [4, 3, 2, 7, 6, 5, 4, 3, 2],
        });
    return String(sum % 10) !== check;
}
const impl = {
    name: 'Bulgarian VAT Number',
    localName: 'Идентификационен номер по ДДС',
    abbreviation: 'ДДС номер',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9 && value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!['0', '1', '2', '3', '9'].includes(value[0])) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length === 9) {
            if (!checkLegal(value)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if (!egn.validate(value).isValid &&
            !pnf.validate(value).isValid &&
            !checkOther(value)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: value.length === 10,
            isCompany: value.length === 9,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
