// stdnum@1.4.10/lib/esm/dk/cpr.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
export function getBirthDate(value) {
    const [dob] = strings.splitAt(value, 6);
    let [day, month, year] = strings.splitAt(dob, 2, 4).map(v => Number(v));
    if ('5678'.includes(value[6]) && year >= 58) {
        year += 1800;
    }
    else if ('0123'.includes(value[6]) ||
        ('49'.includes(value[6]) && year >= 37)) {
        year += 1900;
    }
    else {
        year += 2000;
    }
    const date = new Date(year, month - 1, day);
    if (Number.isNaN(date.getTime()) ||
        [year, String(month).padStart(2, '0'), String(day).padStart(2, '0')].join('-') !== date.toISOString().substr(0, 10)) {
        throw new exceptions.InvalidComponent('The number does not contain valid birth date information.');
    }
    return date;
}
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -');
    if (err !== null) {
        return [value, err];
    }
    return [value, null];
}
const impl = {
    name: 'Danish Citizen Number',
    localName: 'Personnummer',
    abbreviation: 'CPR',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 6).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        try {
            const date = getBirthDate(value);
            if (date.getTime() > new Date().getTime()) {
                return {
                    isValid: false,
                    error: new exceptions.InvalidComponent('The birth date information is valid, but this person has not been born yet.'),
                };
            }
        }
        catch (err) {
            if (err instanceof exceptions.ValidationError) {
                return {
                    isValid: false,
                    error: err,
                };
            }
            return {
                isValid: false,
                error: new exceptions.InvalidComponent(String(err)),
            };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
