// stdnum/lib/cjs/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidComponent = exports.InvalidLength = exports.InvalidChecksum = exports.InvalidFormat = exports.ValidationError = void 0;
class ValidationError extends Error {
    constructor(msg) {
        super(msg);
        this.name = 'ValidationError';
    }
}
exports.ValidationError = ValidationError;
class InvalidFormat extends ValidationError {
    constructor(msg = 'The number has an invlidad format.') {
        super(msg);
        this.name = 'InvalidFormat';
    }
}
exports.InvalidFormat = InvalidFormat;
class InvalidChecksum extends ValidationError {
    constructor(msg = "The number number's checksum or check digit is invalid.") {
        super(msg);
        this.name = 'InvalidChecksum';
    }
}
exports.InvalidChecksum = InvalidChecksum;
class InvalidLength extends ValidationError {
    constructor(msg = 'The number has an invalid length.') {
        super(msg);
        this.name = 'InvalidLength';
    }
}
exports.InvalidLength = InvalidLength;
class InvalidComponent extends ValidationError {
    constructor(msg = 'One of the parts of the number are invalid or unknown.') {
        super(msg);
        this.name = 'InvalidComponent';
    }
}
exports.InvalidComponent = InvalidComponent;
