// stdnum@1.4.10/lib/esm/tr/vkn.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, '');
}
const impl = {
    name: 'Turkish Tax Identification Number',
    localName: 'Vergi Kimlik Numarası',
    abbreviation: 'VKN',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, check] = strings.splitAt(value, -1);
        const sum = front
            .split('')
            .map((v, i) => (parseInt(v, 10) + 9 - i) % 10)
            .map((v, i) => {
            if (v === 0) {
                return v;
            }
            return (v * Math.pow(2, (9 - i))) % 9 || 9;
        })
            .reduce((acc, v) => (acc + v) % 10);
        if (String((10 - sum) % 10) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
