// stdnum/lib/esm/fr/nir.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' .');
}
const impl = {
    name: 'French Personal Identification Number',
    localName: 'Numéro d’Inscription au RNIPP',
    abbreviation: 'NIR',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1, 3, 5, 7, 10, 13).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 15) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [pre, dept, post, check] = strings.splitAt(value, 5, 7, 13);
        if (!strings.isdigits(pre) ||
            !strings.isdigits(post) ||
            !strings.isdigits(check)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        let add = '';
        if (dept === '2A') {
            add = '18';
        }
        else if (dept === '2B') {
            add = '19';
        }
        else if (strings.isdigits(dept)) {
            add = dept;
        }
        else {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = parseInt(`${pre}${add}${post}`, 10) % 97;
        if (String(97 - sum).padStart(2, '0') !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
