// stdnum/lib/cjs/mx/clabe.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const banks_1 = require("./banks");
function clean(input) {
    return util_1.strings.cleanUnicode(input, '- ');
}
const impl = {
    name: 'Mexican Standardized Bank Code',
    localName: 'Clave Bancaria Estandarizada',
    abbreviation: 'CLABE',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 18) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [bankCode, cityCode, account, checksum] = util_1.strings.splitAt(value, 3, 6, 17);
        if (banks_1.banksMap[parseInt(bankCode, 10)] === undefined) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (banks_1.cities[parseInt(cityCode, 10)] === undefined) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!util_1.strings.isdigits(account)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const sum = (10 -
            (0, util_1.weightedSum)(value.substr(0, 17), {
                weights: [3, 7, 1],
                modulus: 10,
            })) %
            10;
        if (checksum !== String(sum)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
