// stdnum@1.4.10/lib/cjs/nl/btw.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const checksum_1 = require("../util/checksum");
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -.', 'NL');
    if (err) {
        return [value, err];
    }
    const [a, b] = util_1.strings.splitAt(value, -3);
    return [`${a.padStart(9, '0')}${b}`, null];
}
const impl = {
    name: 'Dutch VAT Number',
    localName: 'Btw-identificatienummer',
    abbreviation: 'Btw-nr.',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [a, b, c] = util_1.strings.splitAt(value, 9, 10);
        if (!util_1.strings.isdigits(a) || !util_1.strings.isdigits(c)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (b !== 'B') {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const sum = (0, checksum_1.weightedSum)(a, {
            weights: [9, 8, 7, 6, 5, 4, 3, 2, -1],
            modulus: 10000,
        });
        if (sum % 11 !== 0 && !(0, checksum_1.mod97base10Validate)(`NL${a}${b}${c}`)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: `${a}${b}${c}`,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
