// stdnum@1.4.10/lib/esm/py/ruc.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Paraguay Tax Number',
    localName: 'Registro Único del Contribuyentes',
    abbreviation: 'RUC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return `${value.substr(0, value.length - 1)}-${value.substr(value.length - 1)}`;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length < 5 || value.length > 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [front, check] = strings.splitAt(value, value.length - 1);
        const sum = front
            .split('')
            .reverse()
            .map(x => parseInt(x, 10))
            .reduce((acc, digit, idx) => acc + digit * (idx + 2), 0);
        const digit = String((11 - (sum % 11)) % 10);
        if (check !== digit) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: parseInt(front, 10) < 80000000,
            isCompany: front.length === 8 && parseInt(front, 10) > 80000000,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
