// stdnum@1.4.10/lib/cjs/by/unp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const checksum_1 = require("../util/checksum");
const PREFIX = ['УНП', 'УНП', 'UNP', 'UNP'];
const alphabet = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -/');
    if (err !== null) {
        return [value, err];
    }
    const v2 = PREFIX.reduce((acc, p) => acc || (!value.startsWith(p) ? acc : value.substr(p.length)), null);
    return [v2 !== null && v2 !== void 0 ? v2 : value, null];
}
const impl = {
    name: 'Belarus VAT Number',
    localName: 'Учетный номер плательщика',
    abbreviation: 'УНП (UNP)',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value.substr(2))) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!util_1.strings.isdigits(value[0]) && !'1234567ABCEHKM'.includes(value[0])) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!util_1.strings.isdigits(value[1]) && !'ABCEHKMOPT'.includes(value[1])) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [front, check] = util_1.strings.splitAt(value, -1);
        let frontClean = front;
        if (!util_1.strings.isdigits(front.substr(0, 2))) {
            const vv = String('ABCEHKMOPT'.indexOf(front[1]));
            frontClean = `${front[0]}${vv}${front.substr(2)}`;
        }
        const sum = (0, checksum_1.weightedSum)(frontClean, {
            modulus: 11,
            weights: [29, 23, 19, 17, 13, 7, 5, 3],
            alphabet,
        });
        if (sum === 10 || String(sum) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
