// stdnum@1.4.10/lib/esm/es/nif.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import * as nie from './nie';
import * as dni from './dni';
import * as cif from './cif';
function clean(input) {
    const [value, err] = strings.cleanUnicode(input, ' -./');
    if (err) {
        return [value, err];
    }
    if (value.startsWith('ES')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
const impl = {
    name: 'Spanish VAT Number',
    localName: 'Número de Identificación Fiscal',
    abbreviation: 'NIF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if ('KLM'.includes(value[0])) {
            if (value[value.length - 1] !== dni.calcCheckDigit(value.substr(1))) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
            return {
                isValid: true,
                compact: value,
                isIndividual: true,
                isCompany: false,
            };
        }
        if (strings.isdigits(value[0])) {
            return dni.validate(value);
        }
        if ('XYZ'.includes(value[0])) {
            return nie.validate(value);
        }
        return cif.validate(value);
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
