// stdnum/lib/esm/al/nipt.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    let [value, err] = strings.cleanUnicode(input, ' ');
    if (err) {
        return [value, err];
    }
    if (value.startsWith('AL')) {
        value = value.substring(2);
    }
    else if (value.startsWith('(AL)')) {
        value = value.substring(4);
    }
    return [value, null];
}
const impl = {
    name: 'Albanian VAT Number',
    localName: 'Numri i Identifikimit për Personin e Tatueshëm',
    abbreviation: 'NIPT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!/^[JKL]\d{8}[A-Z]$/.test(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
