// stdnum/lib/esm/ad/nrt.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Andorra Tax Register Number',
    localName: 'Número de Registre Tributari',
    abbreviation: 'NRT',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1, 7).join('-');
    },
    validate(value) {
        const [v, error] = clean(value);
        if (error) {
            return { isValid: false, error };
        }
        if (v.length !== 8) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const mid = v.substr(1, v.length - 2);
        if (!strings.isalpha(v[0]) || !strings.isalpha(v[v.length - 1])) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!strings.isdigits(mid)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!'ACDEFGLOPU'.includes(v[0])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (v[0] === 'F' && mid > '699999') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if ('AL'.includes(v[0]) && mid > '699999' && mid < '800000') {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: v,
            isIndividual: 'FE'.includes(v[0]),
            isCompany: !'FE'.includes(v[0]),
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
