// stdnum/lib/cjs/no/fodselsnummer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -:');
}
function checkBirthdate(value) {
    let [dd, mm, yy, rest] = util_1.strings
        .splitAt(value, 2, 4, 6, 9)
        .map(v => parseInt(v, 10));
    if (dd > 40) {
        dd -= 40;
    }
    if (mm > 40) {
        mm -= 40;
    }
    if (rest < 500) {
        yy += 1900;
    }
    else if (rest < 750 && yy > 54) {
        yy += 1800;
    }
    else if (rest < 1000 && yy < 40) {
        yy += 2000;
    }
    else if (rest < 1000 && yy >= 40) {
        yy += 1900;
    }
    else {
        return false;
    }
    return (0, util_1.isValidDate)(String(yy), String(mm), String(dd));
}
const impl = {
    name: 'Norwegian National Identity Number',
    localName: 'Fødselsnummer',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 6).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!checkBirthdate(value)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [front1, check1] = util_1.strings.splitAt(value, 9, 10);
        const [front2, check2] = util_1.strings.splitAt(value, 10);
        const sum1 = (0, util_1.weightedSum)(front1, {
            weights: [3, 7, 6, 1, 8, 9, 4, 5, 2],
            modulus: 11,
        });
        const sum2 = (0, util_1.weightedSum)(front2, {
            weights: [5, 4, 3, 2, 7, 6, 5, 4, 3, 2],
            modulus: 11,
        });
        if (String(11 - sum1) !== check1 || String(11 - sum2) !== check2) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
