// stdnum/lib/cjs/ec/ruc.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const ci = __importStar(require("./ci"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Ecuadorian Company Tax Number',
    localName: 'Registro Único de Contribuyentes',
    abbreviation: 'RUC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 10).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 13) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!ci.validPrefix(value)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (parseInt(value[2], 10) < 6) {
            const [front, end] = util_1.strings.splitAt(value, 10);
            if (end === '000') {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            return ci.validate(front);
        }
        if (value[2] === '6') {
            const [front, end] = util_1.strings.splitAt(value, 9);
            if (end === '0000') {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if ((0, util_1.weightedSum)(front, {
                weights: [3, 2, 7, 6, 5, 4, 3, 2, 1],
                modulus: 11,
            }) !== 0) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if (value[2] === '9') {
            const [front, end] = util_1.strings.splitAt(value, 10);
            if (end === '000') {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if ((0, util_1.weightedSum)(front, {
                weights: [4, 3, 2, 7, 6, 5, 4, 3, 2, 1],
                modulus: 11,
            }) !== 0) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
