// stdnum/lib/esm/cr/cpj.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
const classTypes = {
    '2': ['100', '200', '300', '400'],
    '3': [
        '002',
        '003',
        '004',
        '005',
        '006',
        '007',
        '008',
        '009',
        '010',
        '011',
        '012',
        '013',
        '014',
        '101',
        '102',
        '103',
        '104',
        '105',
        '106',
        '107',
        '108',
        '109',
        '110',
    ],
    '4': ['000'],
    '5': ['001'],
};
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Costa Rica Tax Number',
    localName: 'Cédula de Persona Jurídica',
    abbreviation: 'CPJ',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1, 4).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const items = classTypes[value[0]];
        if (!items) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!items.includes(value.substr(1, 3))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
