// stdnum@1.4.10/lib/cjs/be/insz.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const util_1 = require("../util");
const nn_1 = require("./nn");
const bis_1 = require("./bis");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Belgian Social Security Identification Number',
    localName: 'Identificatienummer van de Sociale Zekerheid',
    abbreviation: 'INSZ, NISS',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const results = [(0, nn_1.validate)(input), (0, bis_1.validate)(input)];
        const validResult = results.find(r => r.isValid);
        if (validResult)
            return validResult;
        const checksumErrorResult = results.find(r => r.error && r.error.name === 'InvalidChecksum');
        return checksumErrorResult || results[0];
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
