// stdnum/lib/esm/pe/cui.js
import * as exceptions from '../exceptions';
import { strings, weightedSum } from '../util';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Peruvian Personal Identification Card',
    localName: 'Cédula Única de Identidad',
    abbreviation: 'CUI',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        if (value.length === 9) {
            return strings.splitAt(value, 8).join('-');
        }
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 8 && value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (value.length === 9) {
            const [front, check] = strings.splitAt(value, 8);
            const sum = weightedSum(front, {
                weights: [3, 2, 7, 6, 5, 4, 3, 2],
                modulus: 11,
            });
            const digits = ['65432110987'[sum], 'KJIHGFEDCBA'[sum]];
            if (!digits.includes(check)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
