// stdnum/lib/esm/za/idnr.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { isValidDateCompactYYMMDD } from '../util/isValidDate';
import { luhnChecksumValidate } from '../util/checksum';
function clean(input) {
    return strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'South African Identity Document Number',
    localName: 'Identity Document Number',
    abbreviation: 'IDNR',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 13) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const [bdate, gender, , citizen] = strings.splitAt(value, 6, 7, 10, 11);
        if (!'01'.includes(citizen)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!strings.isdigits(gender)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!isValidDateCompactYYMMDD(bdate)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!luhnChecksumValidate(value)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
