// stdnum/lib/esm/es/cif.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import { luhnChecksumDigit } from '../util/checksum';
const entityTypes = 'ABCDEFGHJNPQRSUVW';
const letterCheckDigits = 'JABCDEFGHI';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Spanish Company Tax Number',
    localName: 'Código de Identificación Fiscal',
    abbreviation: 'CIF',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1, 8).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [first, body, check] = strings.splitAt(value, 1, 8);
        if (!strings.isdigits(body) ||
            !entityTypes.includes(first) ||
            !(strings.isdigits(check) || letterCheckDigits.includes(check))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const cs = parseInt(luhnChecksumDigit(body), 10);
        const possibleCheckDigits = letterCheckDigits[cs] + String(cs);
        if (!possibleCheckDigits.includes(check)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
