// stdnum/lib/esm/es/nie.js
import * as exceptions from '../exceptions';
import { strings } from '../util';
import * as dni from './dni';
const checkDigits = 'XYZ';
function clean(input) {
    return strings.cleanUnicode(input, ' -');
}
const impl = {
    name: 'Spanish Foreign Number',
    localName: 'Número de Identificación de Extranjero',
    abbreviation: 'NIE',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return strings.splitAt(value, 1, 8).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 9) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [first, body, check] = strings.splitAt(value, 1, 8);
        if (!strings.isdigits(body) ||
            !checkDigits.includes(first) ||
            strings.isdigits(check)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (dni.calcCheckDigit(`${checkDigits.indexOf(first)}${body}`) !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
