// stdnum/lib/cjs/gb/vat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' -/');
    if (err !== null) {
        return [value, err];
    }
    if (value.startsWith('GB')) {
        return [value.substr(2), null];
    }
    return [value, null];
}
const impl = {
    name: 'United Kingdom (and Isle of Man) VAT Number',
    localName: 'Value Added Tax Registration Number',
    abbreviation: 'VAT Reg No',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, 3, 7).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length === 5) {
            if (!util_1.strings.isdigits(value.substr(2))) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            const cvalue = parseInt(value.substr(2), 10);
            if (value.startsWith('GD') && cvalue >= 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (value.startsWith('HA') && cvalue < 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
        }
        else if (value.length === 11 &&
            (value.startsWith('GD8888') || value.startsWith('HA8888'))) {
            if (!util_1.strings.isdigits(value.substr(6))) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            const cvalue = parseInt(value.substr(6, 3), 10);
            if (value.startsWith('GD') && cvalue >= 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (value.startsWith('HA') && cvalue < 500) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            if (cvalue % 97 !== parseInt(value.substr(9, 2), 10)) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else if (value.length === 9 || value.length === 12) {
            if (!util_1.strings.isdigits(value)) {
                return { isValid: false, error: new exceptions.InvalidFormat() };
            }
            const sum = (0, util_1.weightedSum)(value.substr(0, 9), {
                weights: [8, 7, 6, 5, 4, 3, 2, 10, 1],
                modulus: 97,
            });
            if (value[0] === '0') {
                if (![0, 42, 55].includes(sum)) {
                    return { isValid: false, error: new exceptions.InvalidChecksum() };
                }
            }
            else if (sum !== 0) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
