// stdnum@1.4.10/lib/cjs/lv/pvn.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const checksum_1 = require("../util/checksum");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' -', 'LV');
}
const impl = {
    name: 'Latvian VAT Number',
    localName: 'Pievienotās Vērtības Nodokļa',
    abbreviation: 'PVN',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        if ('0123'.includes(value[0])) {
            return util_1.strings.splitAt(value, 6).join('-');
        }
        return util_1.strings.splitAt(value, 4, 8).join(' ');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const isIndividual = '0123'.includes(value[0]);
        if (isIndividual) {
            const [dd, mm, yy, century] = util_1.strings.splitAt(value, 2, 4, 6, 7);
            if (!(0, util_1.isValidDateCompactYYYYMMDD)(`${18 + parseInt(century, 10)}${yy}${mm}${dd}`)) {
                return { isValid: false, error: new exceptions.InvalidComponent() };
            }
            const [front, check] = util_1.strings.splitAt(value, -1);
            const sum = 1 +
                (0, checksum_1.weightedSum)(front, {
                    weights: [10, 5, 8, 4, 2, 1, 6, 3, 7, 9],
                    modulus: 11,
                });
            if (String((sum % 11) % 10) !== check) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        else {
            const sum = (0, checksum_1.weightedSum)(value, {
                weights: [9, 1, 4, 8, 3, 10, 2, 5, 7, 6, 1],
                modulus: 11,
            });
            if (sum !== 3) {
                return { isValid: false, error: new exceptions.InvalidChecksum() };
            }
        }
        return {
            isValid: true,
            compact: value,
            isIndividual,
            isCompany: !isIndividual,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
