// stdnum/lib/cjs/se/personnummer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const checksum_1 = require("../util/checksum");
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, ' :');
    if (err) {
        return [value, err];
    }
    const [a, b, c] = util_1.strings.splitAt(value, -5, -4);
    return [`${a.replace(/[-+]/g, '')}${b}${c}`, null];
}
const impl = {
    name: 'Swedish Personal Identity Number',
    localName: 'Personnummer',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return util_1.strings.splitAt(value, -4).join('-');
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 11 && value.length !== 13) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        const [a, b, c] = util_1.strings.splitAt(value, -5, -4);
        if (!'-+'.includes(b)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        const digits = `${a}${c}`;
        if (!util_1.strings.isdigits(digits)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        let yyyymmdd;
        if (value.length === 13) {
            yyyymmdd = value.substring(0, 8);
        }
        else {
            const yymmdd = value.substring(0, 6);
            const year = new Date().getFullYear();
            let century = Math.floor(year / 100);
            if (parseInt(yymmdd.substring(0, 2), 10) > year % 100) {
                century -= 1;
            }
            if (b === '+') {
                century -= 1;
            }
            yyyymmdd = `${century}${yymmdd}`;
        }
        if (!(0, util_1.isValidDateCompactYYYYMMDD)(yyyymmdd)) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (!(0, checksum_1.luhnChecksumValidate)(digits)) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: true,
            isCompany: false,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
