// stdnum@1.4.10/lib/cjs/ro/onrc.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
const checkRe = /[A-Z]\d+\/\d{1,5}\/\d+/;
function* genit() {
    for (let i = 0; i <= 41; i += 1) {
        yield i;
    }
    yield 51;
    yield 52;
}
const counties = new Set(genit());
function clean(input) {
    const [value, err] = util_1.strings.cleanUnicode(input, '');
    if (err !== null) {
        return [value, err];
    }
    const better = value
        .replace(/[ /\\-]+/g, '/')
        .replace(/^([A-Z])\//, '$1')
        .replace(/([A-Z])(\d)\//, '$10$2')
        .replace(/\/\d{2}[.]\d{2}[.](\d{4})$/, '/$1');
    return [better, null];
}
const impl = {
    name: 'Romanian Trade Register Identifier',
    localName: 'Ordine din Registrul Comerţului',
    abbreviation: 'NRC',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (!checkRe.test(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        if (!'JFC'.includes(value[0])) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        const [county, serial, year] = value.substr(1).split('/');
        if (!counties.has(parseInt(county, 10))) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (serial.length > 5) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        if (year.length !== 4) {
            return { isValid: false, error: new exceptions.InvalidComponent() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: false,
            isCompany: true,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
