// stdnum@1.4.10/lib/cjs/ru/inn.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compact = exports.format = exports.validate = exports.abbreviation = exports.localName = exports.name = void 0;
const exceptions = __importStar(require("../exceptions"));
const util_1 = require("../util");
function clean(input) {
    return util_1.strings.cleanUnicode(input, ' ');
}
const impl = {
    name: 'Russian Tax Identifier',
    localName: 'Идентификационный номер налогоплательщика',
    abbreviation: 'ИНН',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value.toLocaleUpperCase();
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const [value, error] = clean(input);
        if (error) {
            return { isValid: false, error };
        }
        if (value.length !== 10 && value.length !== 12) {
            return { isValid: false, error: new exceptions.InvalidLength() };
        }
        if (!util_1.strings.isdigits(value)) {
            return { isValid: false, error: new exceptions.InvalidFormat() };
        }
        let digit;
        const [front, check] = util_1.strings.splitAt(value, value.length === 10 ? -1 : -2);
        if (value.length === 10) {
            digit = String((0, util_1.weightedSum)(front, {
                weights: [2, 4, 10, 3, 5, 9, 4, 6, 8],
                modulus: 11,
            }) % 10);
        }
        else {
            const d1 = String((0, util_1.weightedSum)(front, {
                weights: [7, 2, 4, 10, 3, 5, 9, 4, 6, 8],
                modulus: 11,
            }) % 10);
            const d2 = String((0, util_1.weightedSum)(front + d1, {
                weights: [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8],
                modulus: 11,
            }) % 10);
            digit = `${d1}${d2}`;
        }
        if (digit !== check) {
            return { isValid: false, error: new exceptions.InvalidChecksum() };
        }
        return {
            isValid: true,
            compact: value,
            isIndividual: value.length === 12,
            isCompany: value.length === 10,
        };
    },
};
exports.name = impl.name, exports.localName = impl.localName, exports.abbreviation = impl.abbreviation, exports.validate = impl.validate, exports.format = impl.format, exports.compact = impl.compact;
