// stdnum/lib/cjs/util/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.strings = exports.isValidDateCompactDDMMYY = exports.isValidDateCompactYYMMDD = exports.isValidDateCompactYYYYMMDD = exports.isValidDate = exports.weightedSum = void 0;
const strings = __importStar(require("./strings"));
exports.strings = strings;
var checksum_1 = require("./checksum");
Object.defineProperty(exports, "weightedSum", { enumerable: true, get: function () { return checksum_1.weightedSum; } });
var isValidDate_1 = require("./isValidDate");
Object.defineProperty(exports, "isValidDate", { enumerable: true, get: function () { return isValidDate_1.isValidDate; } });
Object.defineProperty(exports, "isValidDateCompactYYYYMMDD", { enumerable: true, get: function () { return isValidDate_1.isValidDateCompactYYYYMMDD; } });
Object.defineProperty(exports, "isValidDateCompactYYMMDD", { enumerable: true, get: function () { return isValidDate_1.isValidDateCompactYYMMDD; } });
Object.defineProperty(exports, "isValidDateCompactDDMMYY", { enumerable: true, get: function () { return isValidDate_1.isValidDateCompactDDMMYY; } });
