// stdnum@1.4.10/lib/esm/be/insz.js
import { strings } from '../util';
import { validate as nnValidate } from './nn';
import { validate as bisValidate } from './bis';
function clean(input) {
    return strings.cleanUnicode(input, ' -.');
}
const impl = {
    name: 'Belgian Social Security Identification Number',
    localName: 'Identificatienummer van de Sociale Zekerheid',
    abbreviation: 'INSZ, NISS',
    compact(input) {
        const [value, err] = clean(input);
        if (err) {
            throw err;
        }
        return value;
    },
    format(input) {
        const [value] = clean(input);
        return value;
    },
    validate(input) {
        const results = [nnValidate(input), bisValidate(input)];
        const validResult = results.find(r => r.isValid);
        if (validResult)
            return validResult;
        const checksumErrorResult = results.find(r => r.error && r.error.name === 'InvalidChecksum');
        return checksumErrorResult || results[0];
    },
};
export const { name, localName, abbreviation, validate, format, compact } = impl;
